File: LeetCode/Kai/92ReverseLinkedListII.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = dummy;
        for(int i=0; i<m-1; i++) curr = curr.next;  // curr position: m-1
        for(int i=0; i<n-m; i++) {
            ListNode p = curr;
            for(int j=0; j<n-m-i; j++) p = p.next;   //p position: n-1
            ListNode tmp = p.next.next;
            p.next.next = curr.next;
            curr.next = p.next;
            p.next = tmp;
            
            curr = curr.next;
        }
        return dummy.next;
    }
    
}

File: LeetCode/Liang/92ReverseLinkedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        assert(m >= 1 && m <= n);
        
        // Special case.
        if (m == n) {
            return head;
        }
        
        int step = n - m;
        
        // Traverse to position m
        ListNode *preHead = new ListNode(INT_MIN);
        preHead->next = head;
        ListNode *preFirst = preHead;
        while (m > 1) {
            --m;
            preFirst = preFirst->next;
        }
        
        // Reverse node from [m...n]
        ListNode *last = preFirst->next, *itr = preFirst->next->next;
        while (step > 0 && itr) {
            --step;
            ListNode *tmp = itr->next;
            itr->next = preFirst->next;
            preFirst->next = itr;
            itr = tmp;
        }
        last->next = itr;
        
        // Return
        head = preHead->next;
        delete preHead;
        return head;
    }
};
